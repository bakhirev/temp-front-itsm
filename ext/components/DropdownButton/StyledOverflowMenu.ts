import styled from 'styled-components';

import { OverflowMenu } from '../OverflowMenu';

export const StyledOverflowMenu = styled(OverflowMenu)`
  bottom: 0;
  left: 0;
  position: absolute;
  z-index: -1;
`;
