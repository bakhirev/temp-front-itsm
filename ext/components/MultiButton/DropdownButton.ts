import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';
import { Button } from '../Button';

export const DropdownButton = styled(Button)`
  border-radius: ${({ theme }) => `0 ${theme.borderRadius} ${theme.borderRadius} 0`};
  ${({ kind }) =>
    kind === 'secondary' &&
    css`
      border-left: none;
      padding-left: 1px;
    `}
`;

DropdownButton.defaultProps = {
  theme: DEFAULT_THEME,
};
