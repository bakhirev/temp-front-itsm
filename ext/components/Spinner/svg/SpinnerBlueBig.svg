<svg width="48" height="48" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M43.9553 37.3337C46.0346 34.2217 47.3495 30.6787 47.8128 26.9922C48.0193 25.3483 46.6569 24 45 24C43.3432 24 42.0262 25.3532 41.7532 26.9874C40.3302 35.5067 32.9234 42 24 42C14.0589 42 6.00001 33.9411 6.00001 24C6.00001 15.0766 12.4933 7.6698 21.0126 6.24674C22.6468 5.97377 24 4.65683 24 2.99997C24 1.34312 22.6517 -0.0193537 21.0077 0.187219C17.3213 0.650452 13.7783 1.96536 10.6663 4.0447C6.71955 6.68186 3.64341 10.4301 1.8269 14.8156C0.0103989 19.201 -0.464881 24.0266 0.461164 28.6821C1.38721 33.3377 3.67299 37.6141 7.02945 40.9705C10.3859 44.327 14.6623 46.6128 19.3178 47.5388C23.9734 48.4649 28.799 47.9896 33.1844 46.1731C37.5698 44.3566 41.3181 41.2804 43.9553 37.3337Z" fill="#0062FF"/>
</svg>
