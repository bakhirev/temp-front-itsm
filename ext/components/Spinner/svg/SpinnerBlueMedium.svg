<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M21.9776 18.6669C23.0173 17.1109 23.6748 15.3394 23.9064 13.4962C24.0097 12.6742 23.3284 12 22.5 12C21.6716 12 21.0131 12.6766 20.8766 13.4937C20.1651 17.7534 16.4617 21 12 21C7.02944 21 3.00001 16.9706 3.00001 12C3.00001 7.53833 6.24663 3.83493 10.5063 3.1234C11.3234 2.98691 12 2.32844 12 1.50002C12 0.671589 11.3258 -0.00964632 10.5039 0.0936401C8.66065 0.325257 6.88914 0.982712 5.33316 2.02238C3.35977 3.34096 1.8217 5.2151 0.913451 7.40782C0.00519943 9.60053 -0.232441 12.0133 0.230582 14.3411C0.693605 16.6689 1.83649 18.8071 3.51472 20.4853C5.19295 22.1635 7.33115 23.3064 9.65892 23.7694C11.9867 24.2325 14.3995 23.9948 16.5922 23.0866C18.7849 22.1783 20.6591 20.6402 21.9776 18.6669Z" fill="#0062FF"/>
</svg>
