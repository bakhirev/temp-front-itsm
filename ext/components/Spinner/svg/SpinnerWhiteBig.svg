<svg width="48" height="48" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M43.9553 37.3337C46.0346 34.2217 47.3495 30.6787 47.8128 26.9923C48.0193 25.3483 46.6569 24 45 24C43.3431 24 42.0262 25.3532 41.7532 26.9874C40.3302 35.5068 32.9234 42 24 42C14.0589 42 6 33.9411 6 24C6 15.0766 12.4932 7.66983 21.0126 6.24677C22.6468 5.9738 24 4.65685 24 3C24 1.34315 22.6517 -0.0193258 21.0077 0.187247C17.3213 0.65048 13.7783 1.96539 10.6663 4.04473C6.71953 6.68188 3.64339 10.4302 1.82689 14.8156C0.0103881 19.201 -0.464892 24.0266 0.461153 28.6822C1.3872 33.3377 3.67298 37.6141 7.02944 40.9706C10.3859 44.327 14.6623 46.6128 19.3178 47.5388C23.9734 48.4649 28.799 47.9896 33.1844 46.1731C37.5698 44.3566 41.3181 41.2805 43.9553 37.3337Z" fill="white"/>
</svg>
