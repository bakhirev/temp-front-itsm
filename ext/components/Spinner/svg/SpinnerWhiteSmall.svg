<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M14.6518 12.4446C15.3449 11.4072 15.7832 10.2262 15.9376 8.99742C16.0064 8.44945 15.5523 8 15 8C14.4477 8 14.0087 8.45108 13.9177 8.99581C13.4434 11.8356 10.9745 14 8 14C4.68629 14 2 11.3137 2 8C2 5.02554 4.16441 2.55661 7.00419 2.08226C7.54893 1.99127 8 1.55228 8 1C8 0.447715 7.55055 -0.00644196 7.00258 0.0624156C5.77376 0.216827 4.59275 0.65513 3.55544 1.34824C2.23985 2.22729 1.21446 3.47672 0.608964 4.93853C0.00346269 6.40034 -0.154964 8.00887 0.153718 9.56072C0.4624 11.1126 1.22433 12.538 2.34315 13.6569C3.46197 14.7757 4.88743 15.5376 6.43928 15.8463C7.99113 16.155 9.59966 15.9965 11.0615 15.391C12.5233 14.7855 13.7727 13.7602 14.6518 12.4446Z" fill="white"/>
</svg>
