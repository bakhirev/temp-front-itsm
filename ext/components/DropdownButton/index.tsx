import React, { useEffect, useRef, useState } from 'react';
import OutsideClickHandler from 'react-outside-click-handler';
import type { FC } from 'react';

import { ReactComponent as ChevronDownOutline } from '../Icons/system/ChevronDownOutline.svg';
import { ReactComponent as ChevronUpOutline } from '../Icons/system/ChevronUpOutline.svg';
import { Button } from '../Button';
import { KEY_CODES } from '../common';
import type { IButtonProps } from '../Button';
import type { IOverflowMenuProps } from '../OverflowMenu';

import { Wrapper } from './Wrapper';
import { StyledOverflowMenu } from './StyledOverflowMenu';

const ICON_SIZE = 24;
const ICON_SIZE_SMALL = 20;

type Kind = 'primary' | 'secondary' | 'ghost' | 'white';

export interface IDropdownButtonProps {
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Дефолтное значение меню */
  defaultValue?: IOverflowMenuProps['defaultValue'];
  /** Отключение кнопки */
  disabled?: IButtonProps['disabled'];
  /** Внешний вид кнопки */
  kind?: Kind;
  /** Массив опций */
  list: IOverflowMenuProps['list'];
  /** Размер кнопки */
  size?: IButtonProps['size'];
  /** Значение меню */
  value?: IOverflowMenuProps['value'];
  /** Коллбэк на изменение значения меню. Срабатывает при клике на опцию или нажатии Enter/Space при фокусе на опции */
  onChange?: IOverflowMenuProps['onChange'];
}

export const DropdownButton: FC<IDropdownButtonProps> = ({
  children,
  className,
  dataTestId,
  defaultValue,
  disabled,
  kind,
  list,
  size = 'medium',
  value,
  onChange,
}) => {
  const overflowMenuRef: any = useRef(null);

  const [menuOpened, setMenuOpened] = useState<boolean>(false);

  const handleOutsideClick = () => setMenuOpened(false);
  const handleChange: IOverflowMenuProps['onChange'] = (value) => {
    setMenuOpened(false);
    onChange?.(value);
  };

  const handleMenuOpen = () => {
    setMenuOpened(!menuOpened);
    overflowMenuRef.current?.select.focus();
  };

  const handleEscKeyDown = (event: KeyboardEvent) => {
    if (menuOpened && event.keyCode === KEY_CODES.ESCAPE) {
      setMenuOpened(false);
    }
  };

  useEffect(() => {
    document.addEventListener('keydown', handleEscKeyDown);
    return () => {
      document.removeEventListener('keydown', handleEscKeyDown);
    };
  });

  const iconSize = size === 'big' || size === 'medium' ? ICON_SIZE : ICON_SIZE_SMALL;
  const Icon = menuOpened ? ChevronUpOutline : ChevronDownOutline;

  return (
    <OutsideClickHandler onOutsideClick={handleOutsideClick}>
      <Wrapper className={className} data-test-id={dataTestId}>
        <Button disabled={disabled} kind={kind} size={size} onClick={handleMenuOpen}>
          {children}
          <Icon height={iconSize} width={iconSize} />
        </Button>
        <StyledOverflowMenu
          defaultValue={defaultValue}
          innerRef={overflowMenuRef}
          list={list}
          menu={{
            opened: menuOpened,
          }}
          hideDropdownIndicator
          value={value}
          onChange={handleChange}
        />
      </Wrapper>
    </OutsideClickHandler>
  );
};
