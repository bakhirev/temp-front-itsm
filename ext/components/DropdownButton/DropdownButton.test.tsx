/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable react/display-name */
import React from 'react';
import { shallow } from 'enzyme';

import { ReactComponent as ChevronUpOutline } from '../Icons/system/ChevronUpOutline.svg';
import { ReactComponent as ChevronDownOutline } from '../Icons/system/ChevronDownOutline.svg';
import { IDropdownButtonProps, DropdownButton } from '../DropdownButton';
import type { IButtonProps } from '../Button';
import type { IOverflowMenuProps } from '../OverflowMenu';

jest.mock('../Button', () => ({
  Button: ({ disabled, kind, size, onClick }: IButtonProps) => <button onClick={onClick} />,
}));

jest.mock('../OverflowMenu', () => ({
  OverflowMenu: ({ defaultValue, list, value, onChange }: IOverflowMenuProps) => (
    <div onClick={() => onChange?.(list[2])} />
  ),
}));

jest.mock('../Icons/system/ChevronDownOutline.svg', () => ({
  ReactComponent: () => <div> down icon </div>,
}));

jest.mock('../Icons/system/ChevronUpOutline.svg', () => ({
  ReactComponent: () => <div> up icon </div>,
}));

describe('DropdownButton', () => {
  const requiredProps: IDropdownButtonProps = {
    list: [
      { label: 'foo', value: 1 },
      { label: 'bar', value: 2 },
      { label: 'baz', value: 3 },
    ],
  };

  it('should render component with required props', () => {
    const wrapper = shallow(<DropdownButton {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should pass on button's props to Button component", () => {
    const buttonProps: IDropdownButtonProps = {
      ...requiredProps,
      disabled: true,
      kind: 'white',
      size: 'small',
    };

    const wrapper = shallow(<DropdownButton {...buttonProps} />);
    const button = wrapper.find('Button');

    expect(button.prop('disabled')).toEqual(buttonProps.disabled);
    expect(button.prop('kind')).toEqual(buttonProps.kind);
    expect(button.prop('size')).toEqual(buttonProps.size);
  });

  it("should pass on menu's props to OverflowMenu component", () => {
    const menuProps = {
      defaultValue: { label: 'foo', value: 1 },
      list: [],
      value: { label: 'baz', value: 3 },
    };

    const wrapper = shallow(<DropdownButton {...menuProps} />);
    const styledOverflowMenu = wrapper.childAt(0).childAt(1);

    expect(styledOverflowMenu.prop('list')).toEqual(menuProps.list);
    expect(styledOverflowMenu.prop('defaultValue')).toEqual(menuProps.defaultValue);
    expect(styledOverflowMenu.prop('value')).toEqual(menuProps.value);
  });

  it('should render Button with right icon', () => {
    const wrapper = shallow(<DropdownButton {...requiredProps}>Dropdown</DropdownButton>);

    const button = wrapper.find('Button');
    const children = button.prop('children');
    const lastChild =
      children && Array.isArray(children) ? children[children.length - 1] : children;

    expect(typeof lastChild === 'object').toBeTruthy();
  });

  it('should render Button with arrow down icon by default', () => {
    const wrapper = shallow(<DropdownButton {...requiredProps} />);

    const button = wrapper.find('Button');
    const children = button.prop('children');
    const lastChild =
      children && Array.isArray(children) ? children[children.length - 1] : children;

    expect(lastChild).toEqual(<ChevronDownOutline height={24} width={24} />);
  });

  it('should render Button with arrow up icon when user clicks on it', () => {
    const wrapper = shallow(<DropdownButton {...requiredProps} />);
    const button = wrapper.find('Button');

    button.simulate('click');
    wrapper.update();

    const updatedButton = wrapper.find('Button');
    const children = updatedButton.prop('children');
    const lastChild =
      children && Array.isArray(children) ? children[children.length - 1] : children;

    expect(lastChild).toEqual(<ChevronUpOutline height={24} width={24} />);
  });

  it('should open menu when user clicks on button', () => {
    const wrapper = shallow(<DropdownButton {...requiredProps} />);
    const button = wrapper.find('Button');
    const styledOverflowMenu = wrapper.childAt(0).childAt(1);
    expect(styledOverflowMenu.prop('menu').opened).toBeFalsy();

    button.simulate('click');
    wrapper.update();

    const updatedStyledOverflowMenu = wrapper.childAt(0).childAt(1);
    expect(updatedStyledOverflowMenu.prop('menu').opened).toBeTruthy();
  });

  it('should call onChange with new value when user chooses one the the list', () => {
    const handleChange = jest.fn();
    const wrapper = shallow(<DropdownButton {...requiredProps} onChange={handleChange} />);

    const overflowMenu = wrapper.childAt(0).childAt(1).dive().dive();
    overflowMenu.simulate('click');

    expect(handleChange).toBeCalledWith(requiredProps.list[2]);
  });
});
