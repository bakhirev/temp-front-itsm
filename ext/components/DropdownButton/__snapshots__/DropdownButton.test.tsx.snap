// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownButton should render component with required props 1`] = `
<OutsideClickHandler
  disabled={false}
  display="block"
  onOutsideClick={[Function]}
  useCapture={true}
>
  <styled.div>
    <Button
      onClick={[Function]}
      size="medium"
    >
      <ReactComponent
        height={24}
        width={24}
      />
    </Button>
    <Styled(OverflowMenu)
      hideDropdownIndicator={true}
      innerRef={
        Object {
          "current": null,
        }
      }
      list={
        Array [
          Object {
            "label": "foo",
            "value": 1,
          },
          Object {
            "label": "bar",
            "value": 2,
          },
          Object {
            "label": "baz",
            "value": 3,
          },
        ]
      }
      menu={
        Object {
          "opened": false,
        }
      }
      onChange={[Function]}
    />
  </styled.div>
</OutsideClickHandler>
`;
